File: src/Integra/Solver.hs
module Integra.Solver (solveLinear, solveQuadratic, solveCubic, maybeShowRoots) where

import Data.Complex
import Data.List (intercalate)
import Integra.AST (Expr(..))
import Integra.Evaluator (eval, showComplex)
import Integra.Derive (hasVar)

evalWithX :: Double -> Expr -> Complex Double
evalWithX x e = eval (x :+ 0) (0 :+ 0) e

eps :: Double
eps = 1e-12

solveLinear :: Expr -> String
solveLinear (Eqq l r) = solveLinear (Sub l r)
solveLinear expr
    | not (hasVar expr) = "No variable x found in expression"
    | otherwise =
        let b = evalWithX 0 expr
            a = evalWithX 1 expr - b
        in if magnitude a < eps
           then if magnitude b < eps
                then "Infinite solutions (identity)"
                else "No solution"
           else "x = " ++ showComplex ((-b) / a)

solveQuadratic :: Expr -> String
solveQuadratic (Eqq l r) = solveQuadratic (Sub l r)
solveQuadratic expr
    | not (hasVar expr) = "No variable x found in expression"
    | otherwise =
        let c = evalWithX 0 expr
            b = (evalWithX 1 expr - evalWithX (-1) expr) / 2
            a = evalWithX 1 expr - b - c
        in if magnitude a < eps
           then solveLinear expr
           else let disc = b*b - 4*a*c
                    sqrtDisc = sqrt disc
                    x1 = (-b + sqrtDisc) / (2*a)
                    x2 = (-b - sqrtDisc) / (2*a)
                in if magnitude disc < eps
                   then "x = " ++ showComplex x1 ++ " (repeated)"
                   else "x₁ = " ++ showComplex x1 ++ "\nx₂ = " ++ showComplex x2

solveCubic :: Expr -> String
solveCubic (Eqq l r) = solveCubic (Sub l r)
solveCubic expr
    | not (hasVar expr) = "No variable x found in expression"
    | otherwise =
        let d = evalWithX 0 expr
            f1 = evalWithX 1 expr
            f_1 = evalWithX (-1) expr
            f2 = evalWithX 2 expr
            b = (f1 + f_1) / 2 - d
            ac = (f1 - f_1) / 2
            a = (f2 - 4*b - 2*ac - d) / 6
            c' = ac - a
        in if magnitude a < eps
           then solveQuadratic expr
           else showCubicRoots a b c' d

showCubicRoots :: Complex Double -> Complex Double -> Complex Double -> Complex Double -> String
showCubicRoots a b c d =
    let p  = (3*a*c - b*b) / (3*a*a)
        q  = (2*b*b*b - 9*a*b*c + 27*a*a*d) / (27*a*a*a)
        disc = (q/2)**2 + (p/3)**3
        sqrtDisc = sqrt disc
        u = (-q/2 + sqrtDisc) ** (1/3 :+ 0)
        v = (-q/2 - sqrtDisc) ** (1/3 :+ 0)
        omega = (-0.5) :+ sqrt 3.0 / 2
        omega2 = omega * omega
        offset = b / (3*a)
        x1 = u + v - offset
        x2 = u*omega + v*omega2 - offset
        x3 = u*omega2 + v*omega - offset
    in if magnitude (x1 - x2) < eps && magnitude (x2 - x3) < eps
       then "x = " ++ showComplex x1 ++ " (triple)"
       else if magnitude (x1 - x2) < eps || magnitude (x2 - x3) < eps
            then let repeated = x1
                     single = if magnitude (x1 - x2) < eps then x3 else x1
                 in "x₁ = " ++ showComplex repeated ++ " (repeated)\nx₂ = " ++ showComplex single
            else "x₁ = " ++ showComplex x1 ++ "\nx₂ = " ++ showComplex x2 ++ "\nx₃ = " ++ showComplex x3

maybeShowRoots :: Expr -> Maybe String
maybeShowRoots (Pow (Num a) (Pow (Num n) (Num (-1)))) = maybeShowRoots (Pow (Num a) (Div (Num 1) (Num n)))
maybeShowRoots (Pow (Num a) (Div (Num 1) (Num n)))
    | n > 1 && abs (fromIntegral (round n) - n) < eps =
        let ni = round n :: Int
            r  = abs a ** (1/n)
            theta = if a >= 0 then 0 else pi
            roots = [ showComplex ((r :+ 0) * cis ((theta + 2*pi*fromIntegral k)/n)) | k <- [0..ni-1] ]
        in Just ("All " ++ show ni ++ " roots:\n  " ++ intercalate "\n  " roots)
maybeShowRoots _ = Nothing

File: wasm/src/solver.rs
use crate::ast::AST;
use crate::complex::{C, nice_num};
use crate::evaluator::eval_real;

fn has_var(ast: &AST) -> bool {
    match ast {
        AST::Var => true,
        AST::Num(_) => false,
        AST::Add(a, b) | AST::Sub(a, b) | AST::Mul(a, b) | AST::Div(a, b) | AST::Pow(a, b) => {
            has_var(a) || has_var(b)
        }
        AST::Sin(a) | AST::Cos(a) | AST::Tan(a) | AST::Exp(a) | AST::Log(a) | AST::Sqrt(a)
        | AST::Abs(a) => has_var(a),
        _ => false,
    }
}

fn solve_from_expr(ast: &AST) -> AST {
    match ast {
        AST::Eqq(l, r) => AST::Sub(l.clone(), r.clone()),
        _ => ast.clone(),
    }
}

pub fn solve_linear(ast: &AST) -> String {
    let expr = solve_from_expr(ast);
    if !has_var(&expr) { return "No variable x found in expression".into(); }
    let b = eval_real(&expr, 0.0);
    let a = eval_real(&expr, 1.0) - b;
    if a.abs() < 1e-12 {
        if b.abs() < 1e-12 { return "Infinite solutions (identity)".into(); }
        return "No solution".into();
    }
    format!("x = {}", nice_num(-b / a))
}

pub fn solve_quadratic(ast: &AST) -> String {
    let expr = solve_from_expr(ast);
    if !has_var(&expr) { return "No variable x found in expression".into(); }
    let c = eval_real(&expr, 0.0);
    let b = (eval_real(&expr, 1.0) - eval_real(&expr, -1.0)) / 2.0;
    let a = eval_real(&expr, 1.0) - b - c;
    if a.abs() < 1e-12 { return solve_linear(&expr); }
    let disc = b * b - 4.0 * a * c;
    if disc < 0.0 {
        let real = -b / (2.0 * a);
        let imag = (-disc).sqrt() / (2.0 * a);
        return format!("x = {} ± {}i", nice_num(real), nice_num(imag));
    }
    let sqrt_d = disc.sqrt();
    let x1 = (-b + sqrt_d) / (2.0 * a);
    let x2 = (-b - sqrt_d) / (2.0 * a);
    if disc.abs() < 1e-12 {
        format!("x = {} (repeated)", nice_num(x1))
    } else {
        format!("x₁ = {}\nx₂ = {}", nice_num(x1), nice_num(x2))
    }
}

pub fn solve_cubic(ast: &AST) -> String {
    let expr = solve_from_expr(ast);
    if !has_var(&expr) { return "No variable x found in expression".into(); }
    let d = eval_real(&expr, 0.0);
    let f1 = eval_real(&expr, 1.0);
    let fm1 = eval_real(&expr, -1.0);
    let f2 = eval_real(&expr, 2.0);
    let b = (f1 + fm1) / 2.0 - d;
    let ac = (f1 - fm1) / 2.0;
    let a = (f2 - 4.0 * b - 2.0 * ac - d) / 6.0;
    let c = ac - a;
    if a.abs() < 1e-12 { return solve_quadratic(&expr); }
    let p = (3.0 * a * c - b * b) / (3.0 * a * a);
    let q = (2.0 * b * b * b - 9.0 * a * b * c + 27.0 * a * a * d) / (27.0 * a * a * a);
    let disc = q * q / 4.0 + p * p * p / 27.0;
    let omega = C::new(-0.5, 3.0_f64.sqrt() / 2.0);
    let omega2 = omega * omega;
    let offset = b / (3.0 * a);
    let u = C::from(-q / 2.0) + C::from(disc.sqrt());
    let v = C::from(-q / 2.0) - C::from(disc.sqrt());
    let cu = u.pow(C::from(1.0 / 3.0));
    let cv = v.pow(C::from(1.0 / 3.0));
    let x1 = cu + cv - C::from(offset);
    let x2 = cu * omega + cv * omega2 - C::from(offset);
    let x3 = cu * omega2 + cv * omega - C::from(offset);
    let eps = 1e-10;
    if (x1 - x2).abs() < eps && (x2 - x3).abs() < eps {
        format!("x = {} (triple)", x1)
    } else if (x1 - x2).abs() < eps || (x2 - x3).abs() < eps {
        let repeated = x1;
        let single = if (x1 - x2).abs() < eps { x3 } else { x1 };
        format!("x₁ = {} (repeated)\nx₂ = {}", repeated, single)
    } else {
        format!("x₁ = {}\nx₂ = {}\nx₃ = {}", x1, x2, x3)
    }
}

pub fn maybe_show_roots(ast: &AST) -> Option<String> {
    match ast {
        AST::Pow(a, b) => {
            match (a.as_ref(), b.as_ref()) {
                (AST::Num(a_val), AST::Div(_, _)) | (AST::Num(a_val), AST::Pow(_, _)) => {
                    let n = match b.as_ref() {
                        AST::Div(num, _) => {
                            if let AST::Num(n) = num.as_ref() { *n } else { return None; }
                        }
                        _ => return None,
                    };
                    if n > 1.0 && (n.round() - n).abs() < 1e-12 {
                        let ni = n.round() as i32;
                        let r = a_val.abs().powf(1.0 / n);
                        let theta = if *a_val >= 0.0 { 0.0 } else { std::f64::consts::PI };
                        let roots: Vec<String> = (0..ni)
                            .map(|k| {
                                let angle = (theta + 2.0 * std::f64::consts::PI * k as f64) / n;
                                C::new(r * angle.cos(), r * angle.sin()).to_string()
                            })
                            .collect();
                        return Some(format!("All {} roots:\n  {}", ni, roots.join("\n  ")));
                    }
                    None
                }
                _ => None,
            }
        }
        _ => None,
    }
}
